gcp/src/main/java/org/apache/iceberg/gcp/gcs/GCSFileIO.java

```diff
@@ -242,4 +250,116 @@ private void internalDeleteFiles(Stream<BlobId> blobIdsToDelete) {
+      if (recoverSoftDeletedObject(blobId)) {
+        return true;
+      }
+
+      // fallback to restoring by copying the latest version
+      if (recoverLatestVersion(blobId)) {
+        return true;
+      }
+
+    } catch (IllegalArgumentException e) {
+      LOG.warn("Invalid GCS path format: {}", path, e);
+    }
+
+    return false;
+  }
+
+  /**
+   * Attempts to restore a soft-deleted object.
+   *
+   * <p>Requires {@code storage.objects.restore} permission
+   *
+   * <p>See <a
+   * href="https://cloud.google.com/storage/docs/use-soft-deleted-objects#restore">docs</a>
+   *
+   * @param blobId the blob identifier
+   * @return {@code true} if blob was recovered, {@code false} if not
+   */
+  protected boolean recoverSoftDeletedObject(BlobId blobId) {
+    try {
+      BucketInfo.SoftDeletePolicy policy = client().get(blobId.getBucket()).getSoftDeletePolicy();
+      if (Duration.ofSeconds(0).equals(policy.getRetentionDuration())) {
+        LOG.warn("Soft delete is disabled for {}", blobId.getBucket());
+        return false;
+      }
+
+      Optional<Blob> latestSoftDeletedBlob =
+          client()
+              .list(
+                  blobId.getBucket(),
+                  Storage.BlobListOption.prefix(blobId.getName()),
+                  Storage.BlobListOption.softDeleted(true))
+              .streamAll()
+              .filter(blob -> blob.getName().equals(blobId.getName()))
+              .max(Comparator.comparing(Blob::getSoftDeleteTime));
+
+      if (latestSoftDeletedBlob.isPresent()) {
+        client().restore(latestSoftDeletedBlob.get().getBlobId());
+        LOG.info("Soft delete object restored file {}", blobId);
+        return true;
+      }
+      LOG.warn("No soft deleted object was found");
+
+    } catch (StorageException e) {
+      LOG.warn("Failed to restore", e);
+    }
+
+    return false;
+  }
+
+  /**
+   * Attempts to restore the latest deleted object version.
+   *
+   * <p>See <a href="https://cloud.google.com/storage/docs/using-versioned-objects#restore">docs</a>
+   *
+   * @param blobId the blob identifier
+   * @return {@code true} if blob was recovered, {@code false} if not
+   */
+  protected boolean recoverLatestVersion(BlobId blobId) {
+    try {
+      if (!client().get(blobId.getBucket()).versioningEnabled()) {
+        LOG.warn("Object versioning is disabled for {}", blobId.getBucket());
+        return false;
+      }
+
+      Optional<Blob> latestVersion =
+          client()
+              .list(
+                  blobId.getBucket(),
+                  Storage.BlobListOption.prefix(blobId.getName()),
+                  Storage.BlobListOption.versions(true))
+              .streamAll()
+              .filter(blob -> blob.getName().equals(blobId.getName()))
+              .max(Comparator.comparing(Blob::getUpdateTimeOffsetDateTime));
+
+      if (latestVersion.isPresent() && latestVersion.get().getDeleteTimeOffsetDateTime() != null) {
```

Comment:
Could we embed the `latestVersion.getDeleteTimeOffsetDateTime` check in the filter above? Then it's the latestVersionBeforeDeletion